drivers/media/i2c/irs1125.c

```diff
@@ -0,0 +1,1137 @@
+    if (ret < 0)
+        dev_err(&client->dev, "%s: i2c write error, reg: %x\n",
+                __func__, reg);
+
+    return ret;
+}
+
+static int irs1125_read(struct v4l2_subdev *sd, u16 reg, u16 *val)
+{
+    int ret;
+    unsigned char data_w[2] = { reg >> 8, reg & 0xff };
+    char rdval[2];
+
+    struct i2c_client *client = v4l2_get_subdevdata(sd);
+
+    ret = i2c_master_send(client, data_w, 2);
+    if (ret < 0)
+    {
+        dev_dbg(&client->dev, "%s: i2c write error, reg: %x\n",
+            __func__, reg);
+        return ret;
+    }
+
+    ret = i2c_master_recv(client, rdval, 2);
+    if (ret < 0)
+        dev_err(&client->dev, "%s: i2c read error, reg: %x\n",
+                __func__, reg);
+
+    *val = rdval[1] | (rdval[0]<<8);
+
+    return ret;
+}
+
+static int irs1125_write_array(struct v4l2_subdev *sd,
+                struct regval_list *regs, int array_size)
+{
+    int i, ret;
+
+    for (i = 0; i < array_size; i++)
+    {
+        if (regs[i].addr == 0xFFFF)
+        {
+            msleep(regs[i].data);
+        }
+        else
+        {
+            ret = irs1125_write(sd, regs[i].addr, regs[i].data);
+            if (ret < 0)
+                return ret;
+        }
+    }
+
+    return 0;
+}
+
+static int irs1125_stream_on(struct v4l2_subdev *sd)
+{
+    int ret, cnt;
+    u16 rdval;
+    struct irs1125 *irs1125 = to_state(sd);
+    struct i2c_client *client = v4l2_get_subdevdata(sd);
+    
+    v4l2_ctrl_grab(irs1125->ctrl_numseq, 1);
+    v4l2_ctrl_grab(irs1125->ctrl_modplls, 1);
+
+    ret = irs1125_write(sd, 0xC400, 0x0001);
+    if (ret < 0)
+    {
+        dev_err(&client->dev, "error enabling firmware: %d", ret);
+        return ret;
+    }
+
+    msleep(100);
+
+    return irs1125_write(sd, 0xA87C, 0x0001);
+}
+
+static int irs1125_stream_off(struct v4l2_subdev *sd)
+{
+    int ret, cnt;
+    u16 rdval;
+    struct irs1125 *irs1125 = to_state(sd);
+    struct i2c_client *client = v4l2_get_subdevdata(sd);
+    v4l2_ctrl_grab(irs1125->ctrl_numseq, 0);
+    v4l2_ctrl_grab(irs1125->ctrl_modplls, 0);
+
+    ret = irs1125_write(sd, 0xA87C, 0x0000);
+    if (ret < 0)
+    {
+        dev_err(&client->dev, "error disabling trigger: %d", ret);
+        return ret;
+    }
+
+    msleep(100);
+
+    return irs1125_write(sd, 0xC400, 0x0002);
+}
+
+static int __sensor_init(struct v4l2_subdev *sd)
+{
+    unsigned int cnt;
+    int ret;
+    u16 rdval;
+    struct i2c_client *client = v4l2_get_subdevdata(sd);
+    
+    cnt = 0;
+    while (1)
+    {
+        ret = irs1125_read(sd, 0xC40F, &rdval);
+        if (ret < 0) {
+            dev_err(&client->dev, "read register 0xC40F failed\n");
+            return ret;
+        }
+        if (CHECK_BIT(rdval, 14) == 0)
+            break;
+            
+        if (cnt >= 5) {
+            dev_err(&client->dev, "timeout waiting for 0xC40F\n");
+            return -EAGAIN;
+        }
+            
+        cnt++;
+        msleep(100);
```

Comment:
Magic sleeps and retries feels a little nasty.
What are we actually waiting on here? The sensor has a ready bit that you're polling? Is there not a defined time in which it will initialise within?

Reply:
Yes, you are right. The datasheet gives a maximum of 390µs between releasing the reset pin and having a "ready" in this register. So I just removed this sleep